#!/bin/bash

#PBS -N hipstr
#PBS -l nodes=1:ppn=1
#PBS -l walltime=10:00:00
#PBS -o /projects/ps-gymreklab/helia/ensembl/biallelic/log
#PBS -e /projects/ps-gymreklab/helia/ensembl/biallelic/log
#PBS -A gymreklab-group
#PBS -q hotel
#PBS -V


cd /projects/ps-gymreklab/helia/ensembl/biallelic/validation/LR/

echo $chr

samtools view -bh -d HP:1 NA19030_happlotagged_chr"$chr".bam > NA19030_haplo1_chr"$chr".bam &&
samtools view -bh -d HP:2 NA19030_happlotagged_chr"$chr".bam > NA19030_haplo2_chr"$chr".bam &&

samtools index NA19030_haplo1_chr"$chr".bam &&
samtools index NA19030_haplo2_chr"$chr".bam &&

dir="/projects/ps-gymreklab/helia/long_read_final_experiments/whole_genome" &&

/projects/ps-gymreklab/helia/long_read_final_experiments/packages/HipSTR_changed/HipSTR/HipSTR 	--bams NA19030_haplo1_chr"$chr".bam \
	--fasta "$dir"/metadata/GCA_000001405.15_GRCh38_no_alt_analysis_set.fa \
	--regions "$dir"/metadata/GRCh38.hipstr_reference_pre.bed \
	--str-vcf hipstr/NA19030_haplo1_chr"$chr".vcf.gz \
	--min-reads 5 \
	--use-unpaired \
	--output-filters \
	--max-str-len 1000 \
	--haploid-chrs chr"$chr" \
	--min-sum-qual -1e18 \
	--max-flank-indel 1 \
	--skip-assembly	\
	--def-stutter-model \
	--bam-samps NA19030 --bam-libs NA19030 &&


/projects/ps-gymreklab/helia/long_read_final_experiments/packages/HipSTR_changed/HipSTR/HipSTR  --bams NA19030_haplo2_chr"$chr".bam \
        --fasta "$dir"/metadata/GCA_000001405.15_GRCh38_no_alt_analysis_set.fa \
        --regions "$dir"/metadata/GRCh38.hipstr_reference_pre.bed \
        --str-vcf hipstr/NA19030_haplo2_chr"$chr".vcf.gz \
        --min-reads 5 \
        --use-unpaired \
        --output-filters \
        --max-str-len 1000 \
        --haploid-chrs chr"$chr" \
        --min-sum-qual -1e18 \
        --max-flank-indel 1 \
        --skip-assembly \
        --def-stutter-model \
        --bam-samps NA19030 --bam-libs NA19030 &&

rm NA19030_haplo2_chr"$chr".bam* rm NA19030_haplo1_chr"$chr".bam*
	



